packages/nextjs/hooks/flashbotRecoveryBundle/useGasEstimation.ts

```diff
@@ -20,51 +20,90 @@ export const useGasEstimation = () => {
     }),
   );
   const { showError } = useShowError();
+
+  // Cache for gas estimates and bundle hash
+  const lastBundleHash = useRef<string>("");
+  const lastEstimates = useRef<BigNumber[]>([]);
+  const lastBaseFeeTimestamp = useRef<number>(0);
+  const lastBaseFee = useRef<BigNumber>(BigNumber.from(0));
+
+  // Function to generate a hash for the transaction bundle
+  const getBundleHash = useCallback((txs: RecoveryTx[]): string => {
+    return txs.map(tx => tx.toEstimate?.data || "").join("");
+  }, []);
+
+  const maxBaseFeeInFutureBlock = async () => {
+    const now = Date.now();
+    // Cache base fee for 1 block (~12 seconds)
+    if (now - lastBaseFeeTimestamp.current < 12000 && !lastBaseFee.current.isZero()) {
+      return lastBaseFee.current;
+    }
+
+    const blockNumberNow = await publicClient.getBlockNumber();
+    console.log("@@@@@@ running api calls");
+    const block = await publicClient.getBlock({ blockNumber: blockNumberNow });
+    const maxBaseFee = FlashbotsBundleProvider.getMaxBaseFeeInFutureBlock(BigNumber.from(block.baseFeePerGas), 3);
+
+    lastBaseFee.current = maxBaseFee;
+    lastBaseFeeTimestamp.current = now;
+    return maxBaseFee;
+  };
+
   const estimateTotalGasPrice = async (
     txs: RecoveryTx[],
     deleteTransaction: (id: number) => void,
     modifyTransactions: (txs: RecoveryTx[]) => void,
   ) => {
     try {
+      const currentBundleHash = getBundleHash(txs);
       let estimates: BigNumber[] = [];
-      if (txs.length <= 3 && targetNetwork.network != "sepolia") {
-        // Try to estimate the gas for the entire bundle
-        const bundle = [...txs.map(tx => tx.toEstimate)];
-        // TODO: Add catching so that if the bundle hasn't changed we don't need to call Alchemy again
-        const simulation = await alchemy.transact.simulateExecutionBundle(bundle as DebugTransaction[]);
-        estimates = simulation.map((result, index) => {
-          if (result.calls[0].error) {
-            console.warn(
-              `Following tx will fail when bundle is submitted, so it's removed from the bundle right now. The contract might be a hacky one, and you can try further manipulation via crafting a custom call.`,
-            );
-            console.warn(index, result);
-            deleteTransaction(index);
-            return BigNumber.from("0");
-          }
-          return BigNumber.from(result.calls[0].gasUsed);
-        });
+
+      // Check if we can use cached estimates
+      if (currentBundleHash === lastBundleHash.current && lastEstimates.current.length === txs.length) {
+        estimates = lastEstimates.current;
```

Comment:
This does seem to work, which is great! The simulateExecutionBundle call costs more than normal calls so a huge win.